packages/components/src/utils/hooks/use-focus-outside.js

```diff
@@ -0,0 +1,189 @@
+/* eslint-disable jsdoc/valid-types */
+/**
+ * Returns true if the given element is a button element subject to focus
+ * normalization, or false otherwise.
+ *
+ * @see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
+ *
+ * @param {EventTarget} eventTarget The target from a mouse or touch event.
+ *
+ * @return {eventTarget is FocusNormalizedButton} Whether element is a button.
+ */
+function isFocusNormalizedButton( eventTarget ) {
+	if ( ! ( eventTarget instanceof window.HTMLElement ) ) {
+		return false;
+	}
+	switch ( eventTarget.nodeName ) {
+		case 'A':
+		case 'BUTTON':
+			return true;
+
+		case 'INPUT':
+			return includes(
+				INPUT_BUTTON_TYPES,
+				/** @type {HTMLInputElement} */ ( eventTarget ).type
+			);
+	}
+
+	return false;
+}
+/* eslint-enable jsdoc/valid-types */
+
+/**
+ * @typedef {import('react').SyntheticEvent} SyntheticEvent
+ */
+
+/**
+ * @callback EventCallback
+ * @param {SyntheticEvent} event input related event.
+ */
+
+/**
+ * @typedef FocusOutsideReactElement
+ * @property {EventCallback} handleFocusOutside callback for a focus outside event.
+ */
+
+/**
+ * @typedef {import('react').MutableRefObject<FocusOutsideReactElement | undefined>} FocusOutsideRef
+ */
+
+/**
+ * @typedef {Object} FocusOutsideReturnValue
+ * @property {EventCallback} onFocus      An event handler for focus events.
+ * @property {EventCallback} onBlur       An event handler for blur events.
+ * @property {EventCallback} onMouseDown  An event handler for mouse down events.
+ * @property {EventCallback} onMouseUp    An event handler for mouse up events.
+ * @property {EventCallback} onTouchStart An event handler for touch start events.
+ * @property {EventCallback} onTouchEnd   An event handler for touch end events.
+ */
+
+/**
+ * A react hook that can be used to check whether focus has moved outside the
+ * element the event handlers are bound to.
+ *
+ * @param {EventCallback} onFocusOutside        A callback triggered when focus moves outside
+ *                                              the element the event handlers are bound to.
+ *
+ * @return {FocusOutsideReturnValue} An object containing event handlers. Bind the event handlers
+ *                                   to a wrapping element element to capture when focus moves
+ *                                   outside that element.
+ */
+export default function useFocusOutside( onFocusOutside ) {
+	const preventBlurCheck = useRef( false );
+
+	/**
+	 * @type {import('react').MutableRefObject<number | undefined>}
+	 */
+	const blurCheckTimeoutId = useRef();
+
+	/**
+	 * Cancel a blur check timeout.
+	 */
+	const cancelBlurCheck = useCallback( () => {
+		clearTimeout( blurCheckTimeoutId.current );
+	}, [] );
+
+	// Cancel blur checks on unmount.
+	useEffect( () => {
+		return () => cancelBlurCheck();
+	}, [] );
+
+	// Cancel a blur check if the callback or ref is no longer provided.
```

Comment:
Why is this effect necessary at all?

Reply:
Mentioned here - https://github.com/WordPress/gutenberg/pull/27369#discussion_r539054687